src/bijectors/corr.jl

```diff
@@ -0,0 +1,118 @@
+# See stan doc for parametrization method:
+# https://mc-stan.org/docs/2_23/reference-manual/correlation-matrix-transform-section.html
+# (7/30/2020) their "manageable expression" is wrong...
+
+struct CorrBijector <: Bijector{2} end
+
+function (b::CorrBijector)(x::AbstractMatrix{<:Real})    
+    w = cholesky(x).U + zero(x) # convert to dense matrix
+    r = _link_chol_lkj(w) 
+    return r
+end
+
+(b::CorrBijector)(X::AbstractArray{<:AbstractMatrix{<:Real}}) = map(b, X)
+
+function (ib::Inverse{<:CorrBijector})(y::AbstractMatrix{<:Real})
+    w = _inv_link_chol_lkj(y)
+    return w' * w
+end
+(ib::Inverse{<:CorrBijector})(Y::AbstractArray{<:AbstractMatrix{<:Real}}) = map(ib, Y)
+
+
+function logabsdetjac(::Inverse{CorrBijector}, y::AbstractMatrix{<:Real})
+    @assert size(y, 1) == size(y, 2)
+    K = size(y, 1)
+    
+    left = zero(eltype(y)) # Initial summand may make looping looks weird :(
+    @inbounds for j=2:K, i=1:(j-1)
+        # left += (K-i-1) * log(1 - tanh(y[i, j])^2) # lacks numerically stable
+        left += (K-i-1) * 2 * (log(2) - log(exp(y[i,j]) + exp(-y[i,j])))
+    end
+    
+    right = zero(eltype(y))
+    @inbounds for j=2:K, i=1:(j-1)
+        right += log(cosh(y[i, j])^2)
+    end
+    
+    return left / 2 - right
+end
+function logabsdetjac(b::CorrBijector, X::AbstractMatrix{<:Real})
+    return -logabsdetjac(inv(b),(b(X))) # It may be more efficient if we can use un-contraint value to prevent call of b
+end
+logabsdetjac(b::CorrBijector, X::AbstractArray{<:AbstractMatrix{<:Real}}) = mapvcat(X) do x
+    logabsdetjac(b, x)
+end
+
+
+function _inv_link_chol_lkj(y)
+    @assert size(y, 1) == size(y, 2)
+    K = size(y, 1)
+
+    z = tanh.(y)
+    w = similar(z)
+    
+    w[1,1] = 1
+    @inbounds for j in 1:K
+        w[1, j] = 1
+    end
+
+    @inbounds for j in 1:K
+        for i in j+1:K
+            w[i, j] = 0
+        end
+        for i in 2:j
+            w[i, j] = w[i-1, j] * sqrt(1 - z[i-1, j]^2)
+        end
```

Comment:
The array `z` and the loops below and above are not needed, it seems you could just write
```suggestion
        w[1, j] = 1
        for i in 2:j
            u = w[i-1, j]
            z = tanh(y[i-1, j])
            w[i-1, j] = u * z
            w[i, j] = u * sqrt(1 - z^2)
        end
        for i in (j + 1):K
            w[i, j] = 0
        end
```

Reply:
BTW can't we just return a triangular matrix so we don't have to fill the other elements?

Reply:
It is not needed by ForwardDiff, but it's not free for other 3 reversediff backends since it can be used to avoid an extra tanh call.

Reply:
There's a test to ensure input and output have same type:
https://github.com/TuringLang/Bijectors.jl/blob/b0aaa98f90958a167a0b86c8e8eca9b95502c42d/test/transform.jl#L66
So a zero filling is necessary, a triangular matrix just postpone the filling to another point.

Reply:
> It is not needed by ForwardDiff, but it's not free for other 3 reversediff backends since it can be used to avoid an extra tanh call.

But it's the standard method here that is used for ForwardDiff, for reverse-mode you defined custom adjoints anyways. And even in the reverse-mode case it probably makes sense to move everything into one loop and fill both `z` and `w` there to avoid all the unnecessary computations. Then you can even make `z` a `LowerTriangular` which would probably speed up the computations in the backward pass.

Reply:
> There's a test to ensure input and output have same type:

That seems weird but shouldn't be touched in this PR. @torfjelde Is that a general design choice?